tensorflow/tsl/lib/random/random_distributions.h

```diff
@@ -20,18 +20,59 @@ limitations under the License.
 #include <cmath>
 #include <type_traits>
 
-#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
 #include "tensorflow/tsl/lib/random/philox_random.h"
 #include "tensorflow/tsl/lib/random/random_distributions_utils.h"
 #include "tensorflow/tsl/platform/types.h"
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
 
 namespace tsl {
 namespace random {
 
 // Helper function to convert a 16-bit integer to a half between [0..1).
 PHILOX_DEVICE_INLINE Eigen::half Uint16ToHalf(uint16 x);
+// Helper function to convert a 16-bit integer to a bfloat16 between [1..2).
+PHILOX_DEVICE_INLINE bfloat16 InternalUint16ToBfloat16(uint16 x);
 // Helper function to convert a 16-bit integer to a bfloat16 between [0..1).
-PHILOX_DEVICE_INLINE bfloat16 Uint16ToGfloat16(uint16 x);
+PHILOX_DEVICE_INLINE bfloat16 Uint16ToBfloat16(uint16 x);
+
+// Helper function to format distribution result in vectorization path,
+// it creates Eigen::Tensor and reuses packet feature with SIMD.
+// This function can only work on CPU
+template <class Distribution, class Generator>
+typename Distribution::ResultType VectorizedFormat(
+    Generator* gen, typename Distribution::FormatFunc functor) {
+  typename Generator::ResultType sample;
+  typename Distribution::ResultType result;
+  const int kResultElementCount = Distribution::kResultElementCount;
+  const int inner_count = Generator::kResultElementCount;
+  const int outer_count = kResultElementCount / inner_count;
+  int offset = 0;
+
+  for (int k = 0; k < outer_count; k++) {
+    sample = (*gen)();
+    for (int i = 0; i < inner_count; i++, offset++) {
+      result[offset] = (*functor)(sample[i]);
+    }
+  }
+  // Tail processing if any.
+  // Put the tail condition out of above loop to improve performance:
+  // it will be executed only once and save time on CPU.
+  if (offset < kResultElementCount) {
+    sample = (*gen)();
+    for (int i = 0; offset < kResultElementCount; i++, offset++) {
+      result[offset] = (*functor)(sample[i]);
+    }
+  }
+
+  typedef Eigen::TensorMap<
+      Eigen::Tensor<typename Distribution::ResultElementType, 1,
+                    Eigen::RowMajor, Eigen::DenseIndex>,
+      Eigen::Aligned>
+      Tensor;
+  auto tensor_result = Tensor(&result[0], kResultElementCount);
```

Comment:
Also note that I don't think that the `result` types are aligned.

Reply:
This is still an issue.

Reply:
Thanks for the notice, done.